### PR TITLE
FreeBSD port

### PR DESCRIPTION
**Pull request by kendmerry**
# Summary of changes

This brings in a FreeBSD support for LTFS.

# Description
This brings in FreeBSD support for LTFS.  This requires the FreeBSD tape improvements checked in to FreeBSD in February 2015.  Specifically, it will work on FreeBSD 10.2 or 11.0 or newer releases.

The port was modeled on the lin_tape backend.  It uses a combination of the FreeBSD sa(4) tape driver and the pass(4) SCSI passthrough driver to send commands to the tape drive.

The encryption support has not been tested, but the rest of the functionality has been in production on Spectra Logic Black Pearl products since 2013.  The port also passed LTO Consortium LTFS Compliance Verification:

https://www.lto.org/technology/ltfs/ltfs-compliance-verification/

There is currently no failover support in the FreeBSD backend.  It could be done at some point if someone wants to do it.

This also includes a different mrsw lock implementation for FreeBSD only to resolve some deadlocks that we experienced with the default implementation.  One of my co-workers mentioned it in this issue:

https://github.com/LinearTapeFileSystem/ltfs/issues/8

I am working on a second pull request with most of the rest of Spectra Logic's changes to LTFS.  That one isn't ready yet.  It is mostly independent of this change, but there may be some minor dependencies.

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have confirmed my fix is effective or that my feature works

### CODE REVIEW

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 1]
```diff
@@ -0,0 +1,4206 @@
...
+		goto bailout;
+	}
+
+	timeout = camtape_get_timeout(softc->timeouts, READ_ATTRIBUTE);
+	if (timeout < 0) {
+		rc = -EDEV_UNSUPPORETD_COMMAND;
+		goto bailout;
+	}
+
+
```
Comment (lucasvr) [inline src/tape_drivers/freebsd/cam/camtape_tc.c]: Please see my comments above regarding two consecutive empty lines.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 2]
```diff
@@ -0,0 +1,4206 @@
...
+{
+	int rc;
+	char *msg = NULL;
+	struct camtape_data *softc = device;
+	struct allow_data_overwrite append_pos;
+	int timeout;
+	union ccb *ccb;
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_ENTER(REQ_TC_ALLOWOVERW));
+	ltfsmsg(LTFS_DEBUG, 3059
```
Comment (lucasvr) [inline src/tape_drivers/freebsd/cam/camtape_tc.c]: Interesting, thanks for the detailed comments.

Comment (lucasvr) [inline src/tape_drivers/freebsd/cam/camtape_tc.c]: The indentation looks a bit funny here

Comment (lucasvr) [inline src/tape_drivers/freebsd/cam/camtape_tc.c]: Please see my comments above regarding two consecutive empty lines.

Comment (lucasvr) [inline src/tape_drivers/freebsd/cam/camtape_tc.c]: Two consecutive empty lines

Comment (lucasvr) [inline src/tape_drivers/freebsd/cam/camtape_tc.c]: The tabulation looks a bit funny, at least when rendered by Github. Could you double check (and perhaps change to spaces)?

Comment (lucasvr) [inline src/tape_drivers/freebsd/cam/camtape_tc.c]: Just for the sake of matching the coding style adopted by the project, could you remove this empty line?

Comment (piste-jp-ibm) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: My be github render the code with tab width = 8 by default.

You can see the code fine if you added the "?ts=4" after URL like https://github.com/kendmerry/ltfs/blob/a4f0eed43c93a037e718921e63d1198bc348e556/src/tape_drivers/freebsd/cam/camtape_tc.c?ts=4

**Comment by piste-jp-ibm** [comment]
Hi Ken, 

Thank you for your great enhancement!! 

It's good overall and I can marge the PL if you can modify (or confirm) the following 2 major comments. 

# Mandatory: The magic word of the copyright

Please use the different magic word for copyright registered to both Spectra Logic and IBM and Spectra Logic only. ` OO_Copyright_BEGIN -  OO_Copyright_END` is used copyright registered only to IBM. And the contents will changed automatically by the `replace_copyright.pl` script.

Following 2 magic words are used at this time. 

- `OO_Copyright_BEGIN -  OO_Copyright_END` -> BSD (IBM Only)
- `ZZ_Copyright_BEGIN -  ZZ_Copyright_END` -> IBM Proprietary (Used in IBM internal code)

So how about to define following two?

- `SS_Copyright_BEGIN -  SS_Copyright_END` -> BSD (Spectra Logic Only)
- `OS_Copyright_BEGIN -  OS_Copyright_END` -> BSD (IBM & Spectra Logic)
 
# Optional: Definition of MultiReaderSingleWriter (MRSW)

Downgrading lock logic in the current code is little bit danger, I think.

``` C
//downgrades a write lock to a read lock
static inline void
writetoread_mrsw(MultiReaderSingleWriter *mrsw)
{
	/*
	 * The original intent of this function was to downgrade from write lock
	 * to read lock with higher priority than incoming write
	 * locks. pthread_rwlock doesn't provide this semantic,so we just
	 * release the lock and reacquire a reader lock on it. This demotion is
	 * only used by _ltfs_fsraw_write_data_unlocked() to release the write
	 * lock before returning. If there are pending writers on the volume
	 * lock at this point, they could prevent
	 * _ltfs_fsraw_write_data_unlocked)_ from returning immediately.
	 */
	pthread_rwlock_unlock(mrsw);
	// Someone can cut in here
	pthread_rwlock_rdlock(mrsw);
}
```

As I commented in Issue #8, it is good to add exclusive mutex into the MRSW for FreeBSD. This is optional, you can do it if you want to now. Or you can split out this change into another PL. If you don't have any plan to do, I will work on this and I will unified MRSW for Linux/macOS and MRSW for FreBSD later. Please let me know what is your preference.

My proposal is as follows

## Structures

``` C
typedef struct {
	pthread_mutex_t lock;
} ltfs_mutex_t;

typedef struct {
	pthread_rwlock_t  lock;
} ltfs_rwlock_t;

typedef struct {
	ltfs_mutex_t write_exclusive_mutex;
	ltfs_rwlock_t rwlock;
	uint32_t long_lock;
};
```
 
## Acquire a write lock

1. Acquire write_exclusive_mutex
2. Acquire rwlock on write

## Release a write lock

1. Release rwlock on write
2. Release write_exclusive_mutex

## Acquire a read lock

1. Acquire write_exclusive_mutex
2. Acquire rwlock on read
3. Release write_exclusive_mutex

## Release a read lock

1. Release rwlock on write

**Comment by piste-jp-ibm** [comment]
Once I will close this PL. 

Could you made additional changes I commented "Mandatory" in the previous comment at least? And please create the new PL again when the changes are finished.

Then I will merge this patch to the master branch.

Thank you your contribution in advance !

File: src/libltfs/ltfs.c [hunk 3]
```diff
@@ -200,7 +200,11 @@ int ltfs_set_signal_handlers(void)
 }
 #else
 {
+#ifdef __FreeBSD__
+	sig_t ret;
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/libltfs/ltfs.c]: Good point. It turns out there is already a libltfs/arch/signal_internal with OS-specific ifdefs, so I put the typedef in there.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 4]
```diff
@@ -0,0 +1,4206 @@
...
+			}
+		}
+		i += param_len + LOG_PAGE_PARAM_OFFSET;
+	}
+
+	return -1;
+}
+
+/**
+ * Parse option for IBM tape driver
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 5]
```diff
@@ -0,0 +1,4206 @@
...
+ * @param devname device name of the LTO tape driver
+ * @return a pointer to the camtape backend on success or NULL on error
+ */
+#define IBMTAPE_OPT(templ,offset,value) { templ, offsetof(struct camtape_global_data, offset), value }
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 6]
```diff
@@ -0,0 +1,4206 @@
...
+			camtape_process_errors(device, rc, msg, "write", true);
+
+		if (rc == -EDEV_LBP_WRITE_ERROR)
+			ltfsmsg(LTFS_ERR, 30477E);
+	} else {
+		rc = DEVICE_GOOD;
+		pos->block++;
+	}
+
+	((struct camtape_data *) device)->dirty_acq_loss_w = true;
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 7]
```diff
@@ -0,0 +1,4206 @@
...
+ * Rewind tape
+ * @param device a pointer to the camtape backend
+ * @param pos a pointer to position data. This function will update position infomation.
+ * @return 0 on success or a negative value on error
+ */
+int camtape_rewind(void *device, struct tc_position *pos)
+{
+	int rc;
+	char *msg = NULL;
+	struct camtape_data *softc = device;
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: It means "soft configuration", and is a frequent convention for driver-private variables in FreeBSD and perhaps other BSDs.  I changed it to be consistently "softc" throughout the FreeBSD backend.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 8]
```diff
@@ -0,0 +1,4206 @@
...
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_ENTER(REQ_TC_REWIND));
+	ltfsmsg(LTFS_DEBUG, 30592D, "rewind", ((struct camtape_data *) device)->drive_serial);
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 9]
```diff
@@ -0,0 +1,4206 @@
...
+	char *msg = NULL;
+	struct camtape_data *softc = device;
+	struct scsi_sense_data sense_data;
+	struct mtlocate mtl;
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_ENTER(REQ_TC_LOCATE));
+	ltfsmsg(LTFS_DEBUG, 30597D, "locate", (unsigned long long)dest.partition,
+		(unsigned long long)dest.block, softc->drive_serial);
+
+	bzero(&mtl, sizeof(mtl));
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 10]
```diff
@@ -0,0 +1,4206 @@
...
+{
+	int cmd;
+	int rc;
+	char *msg = NULL;
+	struct camtape_data *softc = device;
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_ENTER(REQ_TC_SPACE));
+	switch (type) {
+		case TC_SPACE_EOD:
+			ltfsmsg(LTFS_DEBUG, 30592D, "space to EOD", ((struct camtape_data *) device)->drive_serial);
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 11]
```diff
@@ -0,0 +1,4206 @@
...
+ * @param type specify type of move
+ * @param pos a pointer to position data. This function will update position infomation.
+ * @return rc 0 on success or a negative value on error
+ */
+int camtape_space(void *device, size_t count, TC_SPACE_TYPE type, struct tc_position *pos)
+{
+	int cmd;
+	int rc;
+	char *msg = NULL;
+	struct camtape_data *softc = device;
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 12]
```diff
@@ -0,0 +1,4206 @@
...
+	CCB_CLEAR_ALL_EXCEPT_HDR(ccb);
+	bzero(&ext_data, sizeof(ext_data));
+
+	timeout = camtape_get_timeout(softc->timeouts, READ_POSITION);
+	if (timeout < 0) {
+		rc = -EDEV_UNSUPPORETD_COMMAND;
+		goto bailout;
+	}
+
+
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

File: src/tape_drivers/freebsd/cam/camtape_tc.c [hunk 13]
```diff
@@ -0,0 +1,4206 @@
...
+	int rc = DEVICE_GOOD;
+	char *msg = NULL;
+	struct camtape_status_item status_items[CT_NUM_STATUS_ITEMS];
+	struct mt_status_data mtinfo;
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_ENTER(REQ_TC_READPOS));
+
+	bzero(status_items, sizeof(status_items));
+	bzero(&mtinfo, sizeof(mtinfo));
+	bcopy(req_status_items, status_items, MIN(sizeof(status_items), sizeof(req_status_items)));
```
[the remark this answers is not in the document]
Comment (kendmerry) [reply src/tape_drivers/freebsd/cam/camtape_tc.c]: Done.

**Comment by kendmerry** [comment]
Hi Abe-san,

I have done all of the above, except for the MRSW changes.  It's fine with me if you do a unified MRSW implementation.

I'll put in another pull request shortly with the updates.  (I don't think I can re-open this one.)

**Comment by piste-jp-ibm** [comment]
Hi Ken, Sure it sounds good to me.

I realized this comment today. I had a notification but I seemed to find wrong places.

I will add an another MRSW implementation which can support Linux, OSX and FreeBSD in the near future.

At first, I will add the switch to use which MRSW architecture. And then, move to new one when we have a confidence of that behavior.

**Comment by kendmerry** [comment]
Hi Abe-san,

Sounds like a good plan.

I have more LTFS changes coming eventually, but I have to balance working on those with working on a number of other unrelated things at the moment.

**Pull request closed without merging**